### PR TITLE
Release GIL around FFTW planning

### PR DESCRIPTION
**Pull request by zpincus**
Instead acquire a python-level lock to keep multiple threads from planning simultaneously. Addresses Issue #75.

### CODE REVIEW

File: pyfftw/pyfftw.pyx [hunk 1]
```diff
@@ -1082,14 +1087,29 @@ cdef class FFTW:
...
+        cdef fftw_iodim *howmany_dims = <fftw_iodim *>self._howmany_dims
+        cdef void *_in = <void *>np.PyArray_DATA(self._input_array)
+        cdef void *_out = <void *>np.PyArray_DATA(self._output_array)
+        cdef int sign = self._direction
+        cdef int c_flags = self._flags
+        
+        plan_lock.acquire()
+        with nogil:
+            plan = fftw_planner(rank, dims, howmany_rank, howmany_dims, 
+                _in, _out, sign, c_flags)
```
Comment (hgomersall) [inline pyfftw/pyfftw.pyx]: It's a weak requirement, but can we make this more pep8y (line up the arguments, unless all on new line).

File: pyfftw/pyfftw.pyx [hunk 2]
```diff
@@ -1082,14 +1087,29 @@ cdef class FFTW:
...
+        plan_lock.release()
+        self._plan = plan
+        
         if self._plan == NULL:
             raise RuntimeError('The data has an uncaught error that led '+
                     'to the planner returning NULL. This is a bug.')
```
Comment (hgomersall) [inline pyfftw/pyfftw.pyx]: This is also not pep8 compliant! (not your problem - legacy code...)

File: pyfftw/pyfftw.pyx (hunk 1 above)
Comment (hgomersall) [inline pyfftw/pyfftw.pyx]: Actually, don't bother with this. I'll go through at some point and make everything pep8 compliant.

File: pyfftw/pyfftw.pyx [hunk 3]
```diff
@@ -1082,14 +1087,29 @@ cdef class FFTW:
...
+        cdef int rank = self._rank
+        cdef fftw_iodim *dims = <fftw_iodim *>self._dims
+        cdef int howmany_rank = self._howmany_rank
+        cdef fftw_iodim *howmany_dims = <fftw_iodim *>self._howmany_dims
+        cdef void *_in = <void *>np.PyArray_DATA(self._input_array)
+        cdef void *_out = <void *>np.PyArray_DATA(self._output_array)
+        cdef int sign = self._direction
+        cdef int c_flags = self._flags
+        
+        plan_lock.acquire()
```
Comment (hgomersall) [inline pyfftw/pyfftw.pyx]: Is there a reason this can't be done with a context manager?

**Comment by zpincus** [comment]
OK, merged in the new master and fixed things to be more pep-8-y and use a context manager for the lock. (I didn't realize you could mix a cython "with nogil" block with normal python context managers, but cython handles this quite nicely.)

Let me know if you'd again prefer a `merge --squash` branch once this is all as desired.

**Comment by hgomersall** [comment]
No, no probs - history is not a problem per se, I just don't like all the crap history that the CI systems seem to result in.

**Comment by zpincus** [comment]
I have no idea why the last commit broke the appveyor build! The previous commit tested fine, but the latest one (which literally only changed a single comment!) is failing on some but not all platforms with:
```
======================================================================
FAIL: test_objects_removed_after_keepalive (test.test_pyfftw_interfaces_cache.CacheTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "C:\projects\pyfftw\test\test_pyfftw_interfaces_cache.py", line 283, in test_objects_removed_after_keepalive
    self.assertRaises(KeyError, _cache.lookup, key)
AssertionError: KeyError not raised by lookup
----------------------------------------------------------------------
```

Any ideas what's going on here?

**Comment by hgomersall** [comment]
Yes, the issue I think is that appveyor builds are scheduled very little time on a virtual machine. The tests that fail depend on a timing threshold not being passed, which it does, because it's just not running to do what it needs to do.

The test should prob be fixed to use cpu time rather than wall time. Though it is actually wall time that is measured.

**Comment by zpincus** [comment]
Ah, OK. Glad to hear my patches didn't somehow cause this!

Let me know if you think this patch-set needs anything else (threaded test-case to demonstrate that the GIL is released during planning? That might be a bit tricky but probably do-able if you think it worthwhile.) or if it looks good to go.

**Comment by hgomersall** [comment]
No, I've no idea how to reliably test threaded release (unless you have any bright ideas). The current failing highlights the problem with testing what is essentially a designed in race condition.

**Comment by zpincus** [comment]
No real bright ideas either that don't basically rely on race conditions. (Actually, any such test would almost by definition be a race condition.)

**Pull request merged**